dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -393,6 +391,321 @@ def save_shm_to_storage(self):
         )
 
 
+class ShardingSaver(CheckpointSaver, ABC):
+    """
+    This saver will save the state dict for all ranks, because the state
+    dict is sharded across all ranks.
+    """
+
+    STAGE_DIR = "._dlrover_ckpt_stage"
+
+    def __init__(self, checkpoint_dir, num_shard=1) -> None:
+        super().__init__(checkpoint_dir, num_shard)
+        self._tensor_shm: List[Optional[SharedMemory]] = [
+            None for _ in range(num_shard)
+        ]
+        self._shared_ckpt_meta = []
+        self._shm_lock = []
+        self._shm_name = []
+
+        # Only local rank 0 will notify the agent to save status to storage
+        qname = _SAVE_STEP_QNAME_PREFIX + str(0)
+        self._to_save_queue = SharedQueue(name=qname, create=True)
+
+        # Each rank has a shared memory to store the state dict
+        for i in range(num_shard):
+            meta_name = _CKPT_META_NAME_PREFIX + str(i)
+            self._shared_ckpt_meta.append(
+                SharedDict(name=meta_name, create=True)
+            )
+            lock_name = _SHM_LOCK_NAME_PREFIX + str(i)
+            self._shm_lock.append(SharedLock(name=lock_name, create=True))
+            shm_name = _TENSOR_SHM_NAME_PREFIX + str(i)
+            self._shm_name.append(shm_name)
+
+        self._node_num = env_utils.get_node_num()
+        self._node_rank = env_utils.get_node_rank()
+        self._is_agent_rank_0 = self._node_rank == 0
+        self._executor = ThreadPoolExecutor(
+            max_workers=self.num_shard, thread_name_prefix="ckpt_saver-"
+        )
+
+        self._writing_storage = False
+
+    @abstractmethod
+    def persist_to_storage(self, state_dict, path):
+        """
+        Persist the checkpoint from CPU memory buffer into the storage.
+        """
+
+    @abstractmethod
+    def update_tracker_file(self, step):
+        pass
+
+    def get_ckpt_name(self, step):
+        """User can override the method to define the checkpoint name."""
+        return f"checkpoint-{step}"
+
+    def _get_ckpt_path(self, step):
```

Comment:
Can we merge line 455 to line 450 to a method?

Reply:
L445 -> L450?